module/events/finalization_actor.go

```diff
@@ -0,0 +1,70 @@
+package events
+
+import (
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/consensus/hotstuff/tracker"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+)
+
+// OnBlockFinalized is invoked when a new block is finalized. It is possible that
+// blocks will be skipped.
+type OnBlockFinalized func(block *model.Block) error
+
+// FinalizationActor is an event responder worker which can be embedded in a component
+// to simplify the plumbing required to respond to block finalization events.
+// This worker is designed to respond to a newly finalized blocks on a best-effort basis,
+// meaning that it may skip blocks when finalization occurs more quickly.
+// CAUTION: This is suitable for use only when the handler can tolerate skipped blocks.
+type FinalizationActor struct {
+	newestFinalized *tracker.NewestBlockTracker
+	notifier        engine.Notifier
+	handler         OnBlockFinalized
+}
+
+// NewFinalizationActor creates a new FinalizationActor, and returns the worker routine
+// and event consumer required to operate it.
+// The caller MUST:
+//   - start the returned component.ComponentWorker function
+//   - subscribe the returned FinalizationActor to OnBlockFinalized events
+func NewFinalizationActor(handler OnBlockFinalized) (*FinalizationActor, component.ComponentWorker) {
+	actor := &FinalizationActor{
+		newestFinalized: tracker.NewNewestBlockTracker(),
+		notifier:        engine.NewNotifier(),
+		handler:         handler,
+	}
+	return actor, actor.worker
+}
+
+// worker is the worker function exposed by the FinalizationActor. It should be
+// attached to a ComponentBuilder by the higher-level component using CreateWorker.
+// It processes each new finalized block by invoking the OnBlockFinalized callback.
+func (actor *FinalizationActor) worker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	doneSignal := ctx.Done()
+	blockFinalizedSignal := actor.notifier.Channel()
+
+	for {
+		select {
+		case <-doneSignal:
+			return
+		case <-blockFinalizedSignal:
+			block := actor.newestFinalized.NewestBlock()
+			err := actor.handler(block)
+			if err != nil {
+				ctx.Throw(err)
+				return
+			}
+		}
+	}
+}
+
+// OnBlockFinalized receives block finalization events. It updates the newest finalized
+// block tracker and notifies the worker thread.
+func (actor *FinalizationActor) OnBlockFinalized(block *model.Block) {
```

Comment:
I have a strong preference to using consistent method names for all the consumers. At the moment, the interface specifies https://github.com/onflow/flow-go/blob/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20/consensus/hotstuff/consumer.go#L33

At the moment, we use anonymous methods for `OnFinalizedBlock` in various places instead of implementing the interface. Therefore, the current code compiles. However, I consider the anonymous methods approach an anti-pattern that I would really like to move away from. Quite frequently, I am wondering which components consume a specific event. But due to the anonymous methods, I can't just search for `OnFinalizedBlock` in the code base. Instead, I need to start at the distributor, figure out that you are routing the events through an anonymous method [here](https://github.com/onflow/flow-go/blob/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20/module/events/finalization_actor.go#L11-L13). So in summary, I realize that the event is processed by the function `handleFinalizedBlock` in the compliance.engine while some other component processes the same event with a function `bar` and another one with a function `foo`. Am I able to describe the challenge?
Overall, I am just missing a clear naming convention, and without a convention there tend to be lots of different inconsistent patters making it super hard to find things.

Generally, I strongly prefer that all consuming functions for an event are named exactly as the interface -- generally the consumer just implements the interface.

Reply:
I've made a few changes:
- In most cases, implement the interface directly
- When subscribing just one function (eg. `AddOnFinalizedBlockConsumer`), use the same function name